src/main/java/cbot/io/Parser.java

```diff
@@ -0,0 +1,362 @@
+            } else if (byIndex + byLength == this.text.length()) {
+                // no due date
+                throw new BadInputException("Missing due date");
+            }
+
+            String dlDesc = this.text.substring(0, byIndex);
+            String dlDueStr = this.text.substring(byIndex + byLength);
+
+            LocalDateTime dlDue = TimeStuff.parseDT(dlDueStr);
+            output = UI.say(tl.addTask(new Deadline(dlDesc, dlDue)));
+            break;
+
+        case EVENT:
+            String fromKeyword = "/from ";
+            String toKeyword = "/to ";
+            int fromLength = fromKeyword.length();
+            int toLength = toKeyword.length();
+
+            if (!this.text.contains(fromKeyword)) {
+                // no /from
+                throw new PoorInputException("Missing \"/from\" keyword");
+            } else if (!this.text.contains(toKeyword)) {
+                // no /to
+                throw new PoorInputException("Missing \"/to\" keyword");
+            }
+
+            int fromIndex = this.text.indexOf(fromKeyword);
+            int toIndex = this.text.indexOf(toKeyword);
+
+            if (toIndex < fromIndex) {
+                // /to before /from
+                throw new PoorInputException("\"/from\" before \"/to\", please!");
+            } else if (fromIndex == 0) {
+                // no desc
+                throw new BadInputException("Missing event description");
+            } else if (fromIndex + fromLength >= toIndex) {
+                // no start
+                throw new BadInputException("Missing start date");
+            } else if (toIndex + toLength == this.text.length()) {
+                // no end
+                throw new BadInputException("Missing end date");
+            }
+
+            String eDesc = this.text.substring(0, fromIndex);
+            String eStartStr = this.text.substring(fromIndex + fromLength, toIndex);
+            String eEndStr = this.text.substring(toIndex + toLength);
+
+            LocalDateTime eStart = TimeStuff.parseDT(eStartStr);
+            LocalDateTime eEnd = TimeStuff.parseDT(eEndStr);
+
+            if (eStart.isAfter(eEnd)) {
+                throw new BadInputException("Hey! You have to start *before* you end...");
+            }
+
+            output = UI.say(tl.addTask(new Event(eDesc, eStart, eEnd)));
+
+            break;
+
+        case SORT:
+            if (tl.getCount() == 0) {
+                throw new PoorInputException("You have no tasks to sort :P");
+            }
+
+            tl.sort();
+            output = UI.say("Okay! I've sorted your tasks by date:\n")
+                    + UI.printMany(tl.listTasks());
+            break;
+
+        case BEFORE:
+            if (tl.getCount() == 0) {
+                throw new PoorInputException("Eh? You have no tasks to filter");
+            }
+
+            LocalDateTime bef = TimeStuff.parseDT(this.text);
+
+            ArrayList<String> arrBef = tl.listFilter(t ->
```

Comment:
This expression seems pretty complicated, perhaps it can be broken down into more intermediate values?